File: aoc23.hs
{-# LANGUAGE TupleSections #-}
{-# OPTIONS_GHC -Wall #-}

import Data.List
import qualified Data.Map as M
import Data.Maybe (fromMaybe, listToMaybe, mapMaybe)
import qualified Data.Set as S
import System.Environment (getArgs)
import qualified Data.Array.IArray as IA
import Data.Array.Unboxed (UArray)

-- import Debug.Trace (trace)

($+$) :: Num a => (a, a) -> (a, a) -> (a, a)
(a, b) $+$ (a', b') = (a + a', b + b')

infixl 6 $+$

rotRight :: Num b => (b, a) -> (a, b)
rotRight (a, b) = (b, -a)

southNbs :: [(Int, Int)]
westNbs :: [(Int, Int)]
northNbs :: [(Int, Int)]
eastNbs :: [(Int, Int)]
southNbs = [(1, -1), (1, 0), (1, 1)]
westNbs = map rotRight southNbs
northNbs = map rotRight westNbs
eastNbs = map rotRight northNbs

allNeighbors :: (Int, Int) -> [(Int, Int)]
allNeighbors (a, b) = [(a + i, b + j) | i <- [-1, 0, 1], j <- [-1, 0, 1], (i /= 0) || (j /= 0)]

doStep :: Int -> S.Set (Int, Int) -> Maybe (S.Set (Int, Int))
doStep tick elves =
    let nxt = foldr S.delete crowded proposedMapRev `S.union` lonely `S.union` M.keysSet proposedMapRev
     in if nxt == elves then Nothing else Just nxt
  where
    arange =
        foldr
            ( \(elfx, elfy) ((minx, miny), (maxx, maxy)) ->
                ((min minx (elfx - 1), min miny (elfy - 1)), (max maxx (elfx + 1), max maxy (elfy + 1)))
            )
            ((0, 0), (0, 0))
            elves
    elfArray = IA.accumArray (const id) True arange [(e, False) | e <- S.toList elves] :: UArray (Int, Int) Bool
    blankspot coord = elfArray IA.! coord
    potdirs = take 4 $ drop (tick `mod` 4) [northNbs, southNbs, westNbs, eastNbs, northNbs, southNbs, westNbs]
    (lonely, crowded) = S.partition (all blankspot . allNeighbors) elves
    proposed elf =
        let allProposed = map (map (elf $+$)) potdirs
            filteredP = filter (all blankspot) allProposed
         in head . tail <$> listToMaybe filteredP
    proposedMapSrc = mapMaybe (\e -> (, Just e) <$> proposed e) $ S.toList crowded
    -- dest -> src
    proposedMapRev = M.mapMaybe id $ M.fromListWith (const $ const Nothing) proposedMapSrc

main :: IO ()
main = do
    args <- getArgs
    let filename =
            if null args
                then "aoc23.in"
                else head args
    s <- lines <$> readFile filename
    let elves = S.fromList $ [(r, c) | (r, line) <- zip [0 ..] s, (c, ch) <- zip [0 ..] line, ch == '#']
    -- part 1
    -- let applyN :: Int -> (a->a) -> a -> a; applyN 0 _ x = x; applyN n f x = applyN (n-1) f (f x)
    let tenSteps = foldl' (\e t -> fromMaybe e (doStep t e)) elves [0 .. 9]
    let spanWidth vs = maximum vs - minimum vs + 1
    let area = spanWidth (S.map fst tenSteps) * spanWidth (S.map snd tenSteps)
    print $ area - S.size tenSteps

    -- part 2
    let p2ans n e = maybe (n + 1) (p2ans $ n + 1) (doStep n e)
    print $ p2ans 0 elves

File: aoc23.py
from aoc_util import *
from collections import Counter
import copy

def elfstep(elves, tick):
    dirtrials = [-1+0j, 1+0j, 0-1j, 0+1j,-1+0j, 1+0j, 0-1j, 0+1j][tick%4:4+(tick%4)]
    elfdest = {}
    for elf in elves:
        neighbors = set([elf + d for d in [-1-1j, -1, -1+1j, 0-1j, 0+1j, 1-1j, 1, 1+1j]])
        if not (neighbors & elves):
            continue
        for d in dirtrials:
            dneigh = set([elf + dp for dp in [(1+1j)*d, d, (1-1j)*d]])
            if not (dneigh & elves):
                elfdest[elf] = elf + d
                break
    dests = Counter(elfdest.values())
    newelves = set()
    done = True
    for elf in elves:
        edest = elfdest.get(elf, elf)
        if dests[edest] > 1:
            edest = elf
        newelves.add(edest)
        if edest != elf:
            done = False

    # minx = int(min(e.real for e in newelves))
    # maxx = int(max(e.real for e in newelves))
    # miny = int(min(e.imag for e in newelves))
    # maxy = int(max(e.imag for e in newelves))

    # for r in range(minx,maxx+1):
    #     for c in range(miny,maxy+1):
    #         if r+c*1j in newelves:
    #             print('#', end='')
    #         else:
    #             print('.', end='')
    #     print()
    # print()

    if done:
        return None
    return newelves

data = chargrid(get_data(23))

elves = set()
for (row, line) in enumerate(data):
    for (col, ch) in enumerate(line):
        if ch == '#':
            elves.add(row + col*1j)
og_elves = copy.deepcopy(elves)

for tick in range(10):
    elves = elfstep(elves, tick)

minx = int(min(e.real for e in elves))
maxx = int(max(e.real for e in elves))
miny = int(min(e.imag for e in elves))
maxy = int(max(e.imag for e in elves))

print((maxx-minx+1)*(maxy-miny+1) - len(elves))

elves = og_elves
tick = 1
while True:
    elves = elfstep(elves, tick-1)
    if elves is None:
        print(tick)
        break
    tick += 1
